homeassistant/components/shopping_list.py

```diff
@@ -256,3 +303,46 @@ def post(self, request):
         hass.data[DOMAIN].async_clear_completed()
         hass.bus.async_fire(EVENT)
         return self.json_message('Cleared completed items.')
+
+
+@websocket_api.async_response
+async def websocket_handle_items(hass, connection, msg):
+    """Handle get shopping_list items."""
+    connection.send_message(websocket_api.result_message(
+        msg['id'], hass.data[DOMAIN].items))
+
+
+@websocket_api.async_response
+async def websocket_handle_add(hass, connection, msg):
+    """Handle add item to shopping_list."""
+    item = hass.data[DOMAIN].async_add(msg['name'])
+    hass.bus.async_fire(EVENT)
+    connection.send_message(websocket_api.result_message(
+        msg['id'], item))
+
+
+@websocket_api.async_response
+async def websocket_handle_update(hass, connection, msg):
+    """Handle update shopping_list item."""
+    try:
+        data = {}
+        if 'name' in msg:
+            data = {'name': msg['name']}
+        else:
+            data = {'complete': msg['complete']}
+        item = hass.data[DOMAIN].async_update(msg['item_id'], data)
+        hass.bus.async_fire(EVENT)
+        connection.send_message(websocket_api.result_message(
+            msg['id'], item))
+    except KeyError:
+        connection.send_message(websocket_api.error_message(
+            msg['id'], 'shopping_list_update_failed', 'Item not found'))
+
+
+@websocket_api.async_response
+async def websocket_handle_clear(hass, connection, msg):
+    """Handle clearing shopping_list items."""
+    hass.data[DOMAIN].async_clear_completed()
+    hass.bus.async_fire(EVENT)
```

Comment:
Isn't it weird that we're duplicating functionality between http and websocket. This should probably be done insidethe `async_clear_completed` method.